Use client side translation for public login page

Chocobo1 opened the pull request:
The translation strings are meant to be synced from Transifex.

sledgehammer999 commented:
Having to specify an id for each translatable element  to grab it via JS, then list all translatable strings in one place in the JS and keep that list in sync with the correct element id will become tedious fast. It is also error prone.

I propose an alternative approach. I didn't find exact documentation for this in i18next itself, but I gathered enough information from examples in stackoverflow and some hints from the i18next-parser.

Take from example the login form:
```html
<form id="loginform">
    <div class="row">
        <label for="username" id="usernameLabel">Username</label><br />
        <input type="text" id="username" name="username" autocomplete="username" required />
    </div>
    <div class="row">
        <label for="password" id="passwordLabel">Password</label><br />
        <input type="password" id="password" name="password" autocomplete="current-password" required />
    </div>
    <div class="row">
        <input type="submit" id="loginButton" value="Login" />
    </div>
</form>
```

Let's drop the `id` and add the strings via a `data-*` attribute, and mark each translatable element with a class `translatable` like so:
```html
<form id="loginform">
    <div class="row">
        <label for="username" data-i18n="Username" class="translatable">Username</label><br />
        <input type="text" id="username" name="username" autocomplete="username" required />
    </div>
    <div class="row">
        <label for="password" data-i18n="Password" class="translatable">Password</label><br />
        <input type="password" id="password" name="password" autocomplete="current-password" required />
    </div>
    <div class="row">
        <input type="submit" data-i18n="Login" class="translatable" value="Login" />
    </div>
</form>
```

Now on the JS side, after the translations are fetched you assign the strings with something like this:
```js
const elements = document.getElementsByClassName("translatable");
  for (let i = 0; i < elements.length; i++) {
    const element = elements[i];
    const k = element.getAttribute("data-i18n");
    element.innerHTML = i18next.t(k);
  }
```

### String extraction into the .json files
If you read the docs of i18next-parser further down where they talk about the options each lexer takes, you'll see that the HTMLLexer supports the `data-i18n` attribute by default and you can add more attributes if you like. So it scan that attribute to extract the "key" (aka the translatable string).

What do you think?

glassez commented:
@Chocobo1 
Is it the intention to apply this translation method to the entire WebUI?

Chocobo1 commented:
>Is it the intention to apply this translation method to the entire WebUI?

Only to the login page. It is written in the PR title.

glassez commented:
> > Is it the intention to apply this translation method to the entire WebUI?
> 
> Only to the login page. It is written in the PR title.
> 
> For now, I have not considered applying to the whole WebUI and I'm not sure how that will turn out.

Is it acceptable to use different translation methods at the same time?

Chocobo1 commented:
>Is it acceptable to use different translation methods at the same time?

It is ok but hopefully it can transition to only one method eventually.

glassez commented:
> > Is it acceptable to use different translation methods at the same time?
> 
> It is ok

Won't this complicate the job of translators?
>but hopefully it can transition to only one method eventually.

Wouldn't it make sense to provide instructions on how to switch the source code to another translation method so that it would be easier for someone else to contribute to such a gradual transition?

Chocobo1 commented:
PR updated. Adopted suggestion from https://github.com/qbittorrent/qBittorrent/pull/20520#issuecomment-1986570164

>Won't this complicate the job of translators?

There will be 2 sources of translation strings of course.

>Wouldn't it make sense to provide instructions on how to switch the source code to another translation method so that it would be easier for someone else to contribute to such a gradual transition?

Would a few basic instructions help? I don't know, it still sounds like a monumental task. There are probably some edge cases and requires specific handling to make it work.

glassez commented on src/webui/www/package.json:
```diff
@@ -6,13 +6,15 @@
     "url": "https://github.com/qbittorrent/qBittorrent.git"
   },
   "scripts": {
+    "extract_translation": "i18next -c public/i18next-parser.config.mjs public/index.html public/scripts/login.js",
```
So, do we need to add each file here separately later?

Chocobo1 replied:
It might support wildcards though.
Also it would be better to append another command to this script: `... && i18next -c private/i18next-parser.config.mjs private/*.html ...`

sledgehammer999 replied:
Yes, it supports globbing patterns as evidence in their readme examples: https://github.com/i18next/i18next-parser

glassez commented:
> Would a few basic instructions help? I don't know, it still sounds like a monumental task. There are probably some edge cases and requires specific handling to make it work.

Or maybe add all the necessary scripts and settings to allow translate files from private folder?

Chocobo1 commented:
>Or maybe add all the necessary scripts and settings to allow translate files from private folder?

You are really interested? I can try lay down the foundations for private folder and leave the string migrations to others.

sledgehammer999 commented on src/webui/www/public/scripts/login.js:
```diff
@@ -62,3 +115,19 @@ function submitLoginForm(event) {
     passwordElement.value = '';
 }
+
+function tr(key) {
```
Why do you introduce the `tr()` function? For familiarity with Qt?
The ` i18next.t()` is an overloaded function, and can return a default value like this:
```js
 i18next.t(key, key);
```

Docs: https://www.i18next.com/translation-function/essentials#passing-a-default-value

sledgehammer999 replied:
Also there is the the following documentation page which I found from google but seems to have been dropped from the navigation menu of the official site (in the past I could navigate to it when I was also investigating i18next):
https://www.i18next.com/principles/fallback#key-fallback

Chocobo1 replied:
>Why do you introduce the tr() function? 
>The  i18next.t() is an overloaded function, and can return a default value like this:

I want it to return the key value when the mapped value is empty and it cannot do that.

Chocobo1 replied:
There is also a second reason related to i18next-parser which will emit a warning when extracting strings. But it is currently work arounded.

sledgehammer999 replied:
> I want it to return the key value when the mapped value is empty (exists but empty, i.e. not yet translated) and `i18next.t()` cannot do that

Are you aware of the `returnEmptyString` config option? Link: https://www.i18next.com/overview/configuration-options#translation-defaults

sledgehammer999 replied:
And I think setting `returnEmptyString = false` will make this `i18next.t(key)` also return the `key` if the it can't find a translated string.

Chocobo1 replied:
>Are you aware of the returnEmptyString config option? Link: https://www.i18next.com/overview/configuration-options#translation-defaults

I wasn't. Good finding!

>And I think setting returnEmptyString = false will make i18next.t(key) also return the key if it can't find a translated string.

True. PR updated.

glassez commented:
> > Or maybe add all the necessary scripts and settings to allow translate files from private folder?
> 
> You are really interested? I can try lay down the foundations for private folder and leave the string migrations to others.

I would just like to "open the doors wider" for other contributors to speed up this process as much as possible.
My motivation is as follows:
I believe that we disabled the applying the translations to alternative WebUIs thoughtlessly, thereby creating a problem for the use of so-called WebUI themes (this is when someone wants to use the original WebUI with minor modifications, mostly visual). Switching the WebUI to client-side translations would automatically save us from this problem. However, if this cannot be done in the near future, I would insist on re-enabling the apllying translations to alternative WebUIs in the upcoming 5.0 version and maybe even in the current v4.6.x branch.

Chocobo1 commented:
>However, if this cannot be done in the near future, I would insist on re-enabling the apllying translations to alternative WebUIs in the upcoming 5.0 version and maybe even in the current v4.6.x branch.

I would prefer there be an option to enable this behavior. (and off by default)

glassez commented on src/webui/www/public/scripts/login.js:
```diff
@@ -28,10 +28,63 @@
+    };
+
+    // present it to i18next
+    const i18nextOptions = {
```
> Since we will download all the languages anyway

Is this really the case? It looks unreasonable. Is it possible to avoid this? Or are we talking about the names of languages, and not about translations? (I'm sorry, I don't go into its details.)

sledgehammer999 replied:
>Is this really the case? It looks unreasonable.

There's a code comment a few lines above. The rationale is it is faster to fetch them all together vs one by one.

IMO, in the case of the "webroot/public/" portion of the WebUI it is ok to do it that way. The translatable strings are very very few.
For the "webroot/private/" portion we would probably need to download them on demand. There is probably a [backend/plugin](https://www.i18next.com/overview/plugins-and-utils#backends) that can help with that. But it is probably best to discuss it in the PR that will implement it for the "webroot/private/" portion.

Chocobo1 replied:
>But we only need single one, don't we?

We can't know if the user preferred translation exists in qbt or not so we fetch all possible matches and present the closest match.
Also fetching them one by one is too slow in case the transmission delay is large.

>Since we will download all the languages anyway, give them all to i18next. Then control the presented langauge via the lng key and the fallback language (aka en) via the fallbackLng key.

IMO there is no point. Feeding more than what i18next needs doesn't improve anything. From my testing it will only use the major language and only when it doesn't exist then the fallback lang will be used.

Chocobo1 replied:
>For the "webroot/private/" portion we would probably need to download them on demand. There is probably a [backend/plugin](https://www.i18next.com/overview/plugins-and-utils#backends) that can help with that. But it is probably best to discuss it in the PR that will implement it for the "webroot/private/" portion.

I would still do it the same way as public folder unless the translation resource is > 1MB large. Besides, the translation resource will be compressed by the server so the actual data over the wire should be way smaller.

sledgehammer999 commented:
>IMO there is no point. Feeding more than what i18next needs doesn't improve anything. From my testing it will only use the major language and only when it doesn't exist then the fallback lang will be used.

I am not sure **currently** how we choose the language to display.
In other websites, the user is able to choose a different language on the fly by clicking a menu item. For that particular case, it makes sense to load everything into i18next and just change the `lgn` key when a user chooses another language. But we probably don't implement that now, so it probably makes sense to load only the specific language.

Chocobo1 commented:
>I am not sure currently how we choose the language to display.

This PR currently detects the preferred language via `navigator.languages` (run it in your browser console to see the lang list). And then the client tries to fetch all possible matches from the server and present the closest/best match.

>In other websites, the user is able to choose a different language on the fly by clicking a menu item. For that particular case, it makes sense to load everything into i18next and just change the lgn key when a user chooses another language. But we probably don't implement that now, so it probably makes sense to load only the specific language.

IMO it would be OK to load the translation resource on demand in this case. I.e user select a language then the client tries to load it instead of loading all translation resources at once at start.
But I imagine this would require more steps when bundling the resources (to build the available lang list) which I'm clueless yet. I would say let's leave the investigation for a later PR.

sledgehammer999 commented:
@Chocobo1 I pushed two commits into your branch, because it was faster for me to implement it than try to explain it here/request it.

The first commit implements a script that will migrate translated strings from the .ts files to the .json files.
It would be wrong to request translators to redo all of their work. This tool can help with the gradual migration of the rest of the WebUI. It updates untranslated .json strings if the translation exists in the corresponding .ts file.

The second commit has pulled the translated strings as they existed before the merging of f265eb0166f57a5dd866a5461b7bf2ab3076ee9f.

If you like, you can merge the second commit into your own. However, I would like the first commit to be as-is unless you want to work on the code/name/whatever more.

sledgehammer999 commented:
>This PR currently detects the preferred language via navigator.languages (run it in your browser console to see the lang list). And then the client tries to fetch all possible matches from the server and present the closest/best match.

@glassez I was wrong. This code doesn't request every single language file. It requests the user's locales (and it's variations) plus the english locale. eg `[el-GR, el, en]`
So, it is ok as it is.

sledgehammer999 commented:
My final objection is in regards of the `.tx/config` file:
Do we really need 2 separate translations resources for the WebUI in Transifex? One for the public part and one for the private.
Can't we workaround this and have only one WebUI resource in Transifex?
Somehow split/merge it locally for the different parts? Or simply duplicate the .json files between the two parts?

sledgehammer999 commented:
>This code doesn't request every single language file. It requests the user's locales (and it's variations) plus the english locale.

@Chocobo1 as an optimization I don't think you need to ever load the `en` JSON, because `.t()` will fallback to the key anyway.

Chocobo1 commented:
>Can't we workaround this and have only one WebUI resource in Transifex? Somehow split/merge it locally for the different parts? 

I honestly don't have better ideas. Do you really insist on it? The public page only has 5 strings so far and is very unlikely to be more in the future.

>Or simply duplicate the .json files between the two parts?

Do you mean exporting json of the private folder and use it for public folder? But won't the translation file become unnecessarily big for the public folder?

Chocobo1 commented:
>@Chocobo1 as an optimization I don't think you need to ever load the en JSON, because .t() will fallback to the key anyway.

I assumed the `navigator.languages` might be empty and when it is, the `lng` parameter for i18next will be undefined. I think that would be weird and decided to still provide `en` as a fallback and make it look consistent.

glassez commented:
> I honestly don't have better ideas. Do you really insist on it? The public page only has 5 strings so far and is very unlikely to be more in the future. The public login page is intentionally kept dead simple.

IMO, the general question is should we keep maintaining this separation of assets of private and public? The really private data is provided via API only.

sledgehammer999 commented:
>I honestly don't have better ideas. Do you really insist on it? The public page only has 5 strings so far and is very unlikely to be more in the future. The public login page is intentionally kept dead simple.

I have slept on it and I propose the following workflow.
Have these 3 locations:
1. `public/lang/<LOCALE>.json`
2. `private/lang/<LOCALE>.json`
3. `transifex/en.json`

The 2 first will contain the generated .json files for all locales for private/public part. Via a script the `en.json` for each part will be merge into one JSON and put under `transifex/`. This will serve as the source file for the relevant single Transifex result.
Now, in reverse, when translations are pulled from Transifex they will be written under `transifex/`. These files are temporary and won't be commited as-is in git. For each locale, a script will load the temporary file and update the corresponding public/private JSON file.

sledgehammer999 commented:
I implemented the new tool and squashed it into my previous commit. Then I force pushed to your branch (preserving your new commit) but the PR hasn't updated.

sledgehammer999 commented on .github/workflows/ci_webui.yaml:
```diff
@@ -38,6 +38,11 @@ jobs:
           npm run format
           git diff --exit-code
 
+      - name: Regenerate translation source
```
Why this step?

Chocobo1 replied:
To make sure the en.json contains the latest up to date strings, for syncing to Transifex.
Let me know if it isn't required now (after your latest changes).

Chocobo1 replied:
Removed.

sledgehammer999 commented on .tx/config:
```diff
@@ -17,6 +17,14 @@ type         = QT
 minimum_perc = 23
 lang_map     = pt: pt_PT, zh: zh_CN
 
+[o:sledgehammer999:p:qbittorrent:r:qbittorrent_webui_public]
+file_filter  = src/webui/www/public/lang/<lang>.json
+source_file  = src/webui/www/public/lang/en.json
```
```
[o:sledgehammer999:p:qbittorrent:r:qbittorrent_webui_json]
file_filter  = src/webui/www/transifex/<lang>.json
source_file  = src/webui/www/transifex/en.json
```

Chocobo1 replied:
>`qbittorrent_webui_json`

isn't it a new resource on transifex? Shouldn't I remove this whole section instead?

> source_file  = src/webui/www/transifex/en.json

Do I also need to change the path that the i18next-parser emits? (in the config file `i18next-parser.config.mjs`)

Chocobo1 replied:
Done.

sledgehammer999 commented on src/webui/www/public/scripts/login.js:
```diff
@@ -28,10 +28,64 @@
+        .map((value, idx) => ({ lang: languages[idx], result: value }))
+        .filter(v => (v.result.value.status === 200));
+    const translation = {
+        lang: translations[0].lang.replace('_', '-'),
```
Check first if `translations` isn't empty.

Chocobo1 replied:
Done.

sledgehammer999 commented on src/webui/www/public/index.html:
```diff
@@ -29,17 +30,17 @@ <h1>qBittorrent WebUI</h1>
             <img src="images/qbittorrent-tray.svg" alt="qBittorrent logo" />
         </div>
         <div id="formplace" class="col">
-            <form id="loginform" method="post" onsubmit="submitLoginForm(event);">
```
Why delete this and use addEventListener instead?

Chocobo1 replied:
IMO it is cleaner to move code things to js instead of mixing html with js.

sledgehammer999 commented on src/webui/www/public/scripts/login.js:
```diff
@@ -62,3 +116,14 @@ function submitLoginForm(event) {
+    loginForm.addEventListener('submit', submitLoginForm);
+
+    document.getElementById('username').focus();
+    document.getElementById('username').select();
```
Maybe instead set the [autofocus](https://www.w3schools.com/tags/att_input_autofocus.asp) attribute on the `username` input?

Chocobo1 replied:
Done.

sledgehammer999 commented:
>I implemented the new tool and squashed it into my previous commit. Then I force pushed to your branch (preserving your new commit) but the PR hasn't updated.

If I somehow broke Github, then you are free to drop my commits and resubmit your own changes (and address the rest of the review comments). Then I'll reintroduce my commits in a new PR.

Chocobo1 commented:
>If I somehow broke Github, then you are free to drop my commits and resubmit your own changes (and address the rest of the review comments). Then I'll reintroduce my commits in a new PR.

It looks OK to me. I can see the 2 python scripts.

sledgehammer999 commented:
@Chocobo1 let me explain my intended workflow which makes the CI change unneeded.

On some interval, I'll pull the translations from Transifex. These will be dropped into `transifex/*.json`.
Then I'll call `split_merge_json.py --mode split` which will split `transifex/*.json` into their equivalent `public/lang/*.json` and `private/lang/*.json` parts. The non english `transifex/*.json` files won't be commited to git.
Afterwards I'll run the `i18next-parser` commands to extract possible new strings into `public/lang/*.json` and `private/lang/*.json`. These will be committed to git.
Finally, I'll run `split_merge_json.py --mode merge` which will merge `public/lang/en.json` and `private/lang/en.json` into `transifex/en.json`. This will be committed to git.
Transifex will be setup to pull regularly `transifex/en.json` from the repo.

> > `qbittorrent_webui_json`
> 
> isn't it a new resource on transifex? Shouldn't I remove this whole section instead?
> 
> > source_file  = src/webui/www/transifex/en.json
> 
> Do I also need to change the path that the i18next-parser emits? (in the config file `i18next-parser.config.mjs`)

So now we need just one resource on Transifex. Since the old WebUI translations will still be active we can't reuse the same name for the resource.

sledgehammer999 commented:
>let me explain my intended workflow which makes the CI change unneeded.

And to add to the above:
When the incremental transition will be made in the `private/` part the author shouldn't generate any *.json files. Just act on the html/js files. The files will be generated by my above workflow. At the appropriate step of my workflow I'll insert a call to `translations_qt2i18next.py` to migrate the strings.

Chocobo1 commented:
@sledgehammer999 
I've updated the PR and should addressed all remaining comments.
I also added `sort_keys=True` to `json.dump()` in the new scripts. This is make sure the output is always sorted and is stable.

sledgehammer999 commented:
I'll probably merge this in the next hours so it can be included in the unstable build.
